source/framework/core/src/TRestDataSet.cxx

```diff
@@ -917,3 +926,66 @@ void TRestDataSet::Import(std::vector<std::string> fileNames) {
 
     fQuantity.clear();
 }
+
+///////////////////////////////////////////////
+/// \brief This function calculates a corrected time of a given Run,
+/// corresponding to the time where a certain variable has remained above a
+/// given threshold (for example, the time for which the detection rate is above a certain value)
+Double_t TRestDataSet::GetRunDuration(TRestRun& r) {
+    Double_t runTime = 0;
+    Double_t corrected_time = 0;
+    if (fTimeCorrection) {
+        Int_t max_events_below_threshold = 5;
+        Int_t events_below_threshold =
+            0;                      // Number of events in a row (except recovery) below the desired threshold
+        Int_t recovery_events = 0;  // Events above the threshold after a series of bad events
+        Int_t first_bad_event = 0;
+        Int_t last_bad_event = 0;
+        Int_t t1 = 0;
+        Int_t t2 = 0;
+        Double_t val = 0.0;
+        if (r.GetAnalysisTree()->GetObservableType(fColumnTimeCorrection) != "double") {
+            RESTError << fColumnTimeCorrection << "is not a double" << RESTendl;
+            exit(1);
+        }
+        for (int ev = 0; ev < r.GetEntries(); ev++) {
+            r.GetEntry(ev);
+            val = r.GetAnalysisTree()->GetObservableValue<double>(fColumnTimeCorrection);
+
+            if (val < fThresholdTimeCorrection) {
+                recovery_events = 0;
+                if (events_below_threshold == 0) {
+                    first_bad_event = ev;
+                }
+                last_bad_event = ev;
+                events_below_threshold++;
+
+            } else {
+                if (events_below_threshold <= max_events_below_threshold) {
+                    events_below_threshold = 0;
+                } else {
+                    recovery_events++;
+                    if (recovery_events > max_events_below_threshold) {
+                        r.GetEntry(first_bad_event);
+                        t1 = r.GetAnalysisTree()->GetObservableValue<double>("rateAna_SecondsFromStart");
+                        r.GetEntry(last_bad_event);
+                        t2 = r.GetAnalysisTree()->GetObservableValue<double>("rateAna_SecondsFromStart");
+                        corrected_time += t2 - t1;
+                        events_below_threshold = 0;
+                    }
+                }
+            }
+            if ((ev == r.GetEntries() - max_events_below_threshold) &&
+                (events_below_threshold > max_events_below_threshold)) {
+                last_bad_event = r.GetEntries() - 1;
+                r.GetEntry(first_bad_event);
+                t1 = r.GetAnalysisTree()->GetObservableValue<double>("rateAna_SecondsFromStart");
```

Comment:
This observable might not be in the analysis tree if the process `rateAna` was not used in the data processing., You should use `timeStamp`.